teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,612 @@
+
+## Requirements
+<!--
+List the requirements for this TEP.
+-->
+1. Users need a way to specify complex `Conditions`, such as those needing access to files in a `Workspace`. 
+1. Users need a way to specify simple `Conditions`, such as checking the name of a git branch matches.
+1. Users need a way to specify what happens when `Conditions` evaluate to `False`.
+
+## Proposal
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+For _simplicity_, we propose deprecating the separate `Conditions` CRD and use `Tasks` instead for guarded execution of other `Tasks`. For _efficiency_, we propose using string expressions through Kubernetes `MatchExpressions` and Tekton’s `CustomTasks` to perform simple checks for guarded execution of `Tasks` without spinning up new `Pods`. For _skipping_, we propose adding a boolean to allow users to specify whether to skip the guarded `Task` only or to skip the guarded `Task` and its dependent `Tasks`. 
+
+This design will give us the flexibility to handle both simple `Guards` (e.g. checking a branch name) using string expressions and complex `Guards` (e.g. those that need a `Workspace` and/or `Volume` with the files to execute on) in a `Pod`, while providing more options when a `Guard` evaluates to `False`. 
+
+
+### Simplicity
+
+As discussed in the background section, `Conditions` manifest themselves as `Tasks`. We want to keep Tekton as simple as possible by reusing existing components. So we propose phasing out the separate `Conditions` CRD and eventually deprecating it. In place of `Conditions`, we want to use `Tasks` - referred to as `Guards` - to determine whether or not a `Task` executes. We refer to them as `Guards` because they determine if a `Task` executes, not if/else as would be expected from a `Condition`; more details on `Guards` vs `Conditions` naming can be found in [this issue](https://github.com/tektoncd/pipeline/issues/2635). Thus, we won’t have `Conditions` to migrate to beta and won’t have to maintain the separate `Conditions` CRD. Moreover, the features supported by `Tasks` become readily available to be used for guarded execution of `Tasks`. 
+
+A `Task` that is used as a `Guard` will produce a `Result` called `Skip` with values `True` or `False` -- to indicate whether the `Task` it’s guarding will be executed or not. When we have more than one `Guard`, the `Task` will be executed when all the `Guards` evaluate to `True`. Using `Results` allows us to distinguish between when a `Task` fails and when the `Guard` evaluates to `False`. 
+
+In the example of checking whether a file exists, the `Task` that would replace the `Condition` would be specified in a list of `when-expressions` as such:
+
```

Comment:
Not a big thing.

But I still feels that it will be a bit complicated/problematic/confusing adding `taskRef:` within this place in the pipeline.

Now these Tasks are proper Tasks - that can also be added to other places in the Pipeline.

Consider a use case: Check if workspace contains Dockerfile - then execute TaskA, if not - excecute TaskB.

That can now be implemented with those _two_ pipelines with **identical result**

```
kind: Pipeline
metadata: 
    name: conditional-dockerfile
spec: 
    tasks:  
      - name: build-w-dockerfile 
        when:
           - name: dockerfile-exist
             taskRef: 
                 name: file-exists
        taskRef:
            name: build-w-dockerfile
      - name: build-wo-dockerfile 
        when:
           - name: dockerfile-n-exist
             taskRef: 
                 name: file-n-exists
        taskRef:
           name: build-wo-dockerfile
```

and - only adding `matchExpressions`

```
kind: Pipeline
metadata: 
    name: conditional-dockerfile-2
spec: 
    tasks: 
      - name: check-file-exists 
        taskRef:
            name: file-exists
      - name: build-w-dockerfile 
        when:
           - key: $(result.check-file-e) 
             operator: isTrue
        taskRef:
            name: build-w-dockerfile
      - name: build-wo-dockerfile 
        when:
           - key: $(result.check-file-e)        
             operator: notTrue
        taskRef:
            name: build-wo-dockerfile
```

The proposal adds both alternatives.

The latter example is **much** easier to implement - we only need to add `matchExpressions` for that - since Tasks and CustomTasks already can be executed within a Pipeline.

The latter example - only need a _single_ "check": `file-exits` while the first example also need `file-n-exists` (file not exists) in addition.

In my opinion - the latter example is also easier to _read_ and _reason_ about.

Reply:
We could perhaps - start with implementing a `when` with `matchExpressions` and if we are not satisfied - continue and add `taskRef:` ?

Reply:
I now added `workspaces` to the examples above. This makes the first example look even more confusing for me - mostly because I have multiple `taskRef` and also multiple `workspaces`-mappings within the _pipelineTask_

Reply:
I agree that given that the `matchExpressions` can operate on `Results` and covers the use cases, we can go with that approach for now and revisit the `Task` with `Skip Result` contract later if needed. 

Thank you for these examples!